### PR TITLE
Cleanup for inactive games - checks every 8 hours, cleans up anything…

### LINKED ISSUE #25: Auto-run !endgame on a town after N hours of inactivity
If no commands come in to a town after like 6-12 hours, and the bot still thinks a game is in progress for that town, run the !endgame command to clean up roles.

### PR DESCRIPTION
**Pull request by magroader**
… with no activity in the last 3 hours

Fixes #25

### CODE REVIEW

File: bot.py [hunk 1]
```diff
@@ -83,7 +83,7 @@ def getTownInfo(self, ctx):
         doc = guildInfo.find_one(query)
 
         if doc:
-            return TownInfo(ctx, doc)
+            return TownInfo(ctx.guild, doc)
```
Comment (lilserf) [inline bot.py:86]: This is gonna fail - ctx doesn't exist

File: bot.py [hunk 2]
```diff
@@ -47,8 +48,7 @@ def getRoleByName(guild, name):
 
 # Nicer than a dict for storing info about the town
 class TownInfo:
-    def __init__(self, ctx, document):
-        guild = ctx.guild
+    def __init__(self, guild, document):
```
Comment (lilserf) [inline bot.py:51]: why swap out ctx for just a guild alone?

File: bot.py [hunk 3]
```diff
@@ -558,6 +558,11 @@ async def sendEmbed(self, ctx, townInfo):
 class Gameplay(commands.Cog):
     def __init__(self, bot):
         self.bot = bot
+        self.currentlyActiveGames = {}
+        #self.cleanupInactiveGames.start() This should only be run if we have active games in the DB
+
+    def cog_unload(self):
```
Comment (lilserf) [inline bot.py:564]: I don't know if this will run in normal operation - when heroku redeploys I doubt it stops cleanly

File: bot.py [hunk 4]
```diff
@@ -873,6 +906,64 @@ async def onVote(self, ctx):
...
+                self.currentlyActiveGames[g].pop(t, None)
+            if not self.currentlyActiveGames[g]:
+                self.currentlyActiveGames.pop(g, None)
+
+        # If nothing is left active at all, stop the timer
+        if not self.currentlyActiveGames:
+            print("No remaining active games, stopping checks")
+            self.cleanupInactiveGames.stop()
+    
+    @cleanupInactiveGames.before_loop
```
Comment (lilserf) [inline bot.py:962]: whoa what is this?

File: bot.py [hunk 5]
```diff
@@ -558,6 +558,11 @@ async def sendEmbed(self, ctx, townInfo):
 class Gameplay(commands.Cog):
     def __init__(self, bot):
         self.bot = bot
+        self.currentlyActiveGames = {}
+        #self.cleanupInactiveGames.start() This should only be run if we have active games in the DB
```
Comment (lilserf) [inline bot.py:562]: Currently without this line the timer only gets started if you call !endgame which feels backwards

**Review by lilserf — CHANGES_REQUESTED** [review]

File: bot.py (hunk 1 above)
Comment (magroader) [reply bot.py:86]: I don't understand - ctx was what was previously being passed here, now it's ctx.guild

File: bot.py (hunk 2 above)
Comment (magroader) [reply bot.py:51]: ctx is not actually required. I need to call this from the timer function which does not have a ctx

File: bot.py (hunk 3 above)
Comment (magroader) [reply bot.py:564]: It was just part of the recommended way to set up the timer so that's what I did

File: bot.py (hunk 4 above)
Comment (magroader) [reply bot.py:962]: It's some sort of attachment for timer functions - this gets called before the timer function is triggered, in this case to make sure we're actually connected to discord

File: bot.py (hunk 5 above)
Comment (magroader) [reply bot.py:562]: It should only get called when you call currgame, not endgame.

I tested by calling ~night and waiting (with shorter timeouts for debugging). It correctly called the timer function in response to ~night.

File: bot.py (hunk 1 above)
Comment (magroader) [reply bot.py:86]: It's possible the phone app is not showing me everything - I'll have to check later on my PC.

File: bot.py (hunk 2 above)
Comment (magroader) [reply bot.py:51]: I need to call this from the timer function because the timer function calls endGameInternal.

File: bot.py (hunk 4 above)
Comment (magroader) [reply bot.py:962]: Apparently functions in python can have properties attached? I guess the timer system looks for this property on the function and if it exists it calls it first.

File: bot.py (hunk 1 above)
Comment (lilserf) [reply bot.py:86]: I missed something on phone - I thought this was the body of the func above where ctx had been removed in favor of guild 😆 didn’t notice the “x lines collapsed” thingy

**Pull request merged**